rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/terminal/TerminalIDEServer.java

```diff
@@ -0,0 +1,60 @@
+package org.rascalmpl.vscode.lsp.terminal;
+
+import java.net.URISyntaxException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.eclipse.lsp4j.MessageParams;
+import org.eclipse.lsp4j.MessageType;
+import org.eclipse.lsp4j.WorkspaceFolder;
+import org.rascalmpl.uri.URIUtil;
+import org.rascalmpl.vscode.lsp.IRascalLanguageClient;
+
+import io.usethesource.vallang.ISourceLocation;
+
+/**
+ * This server forwards IDE services requests by a Rascal terminal
+ * directly to the LSP language client.
+ */
+public class TerminalIDEServer implements ITerminalIDEServer {
+    private static final Logger logger = LogManager.getLogger(TerminalIDEServer.class);
+
+    private final IRascalLanguageClient languageClient;
+
+    public TerminalIDEServer(IRascalLanguageClient client) {
+        this.languageClient = client;
+    }
+
+    @Override
+    public CompletableFuture<Void> browse(BrowseParameter uri) {
+        languageClient.showContent(uri);
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public CompletableFuture<Void> edit(EditParameter edit) {
+        languageClient.showMessage(new MessageParams(MessageType.Info, "trying to edit: " + edit.getModule()));
+        return CompletableFuture.completedFuture(null);
+    }
+
+    @Override
+    public CompletableFuture<ISourceLocation> resolveProjectLocation(ISourceLocation input) {
+        try {
+            for (WorkspaceFolder folder : languageClient.workspaceFolders().get()) {
+                // TODO check if everything goes ok encoding-wise
+                if (folder.getName().equals(input.getAuthority())) {
+                    ISourceLocation root = URIUtil.createFromURI(folder.getUri());
+                    return CompletableFuture.completedFuture(URIUtil.getChildLocation(root, input.getPath()));
+                }
+                
+            }
+            return CompletableFuture.completedFuture(null);
+        }  
+        catch (URISyntaxException | InterruptedException | ExecutionException e) {
+            logger.error(e);
+            return CompletableFuture.completedFuture(input);
```

Comment:
shouldn't this be an completed exceptionally? I don't know how this translates to the call on the other side though.